File: src/game/recipes/bookView.ts
import type { GameDatabase } from '../data/types'
import { getRecipe } from '../production/recipes'
import { getProject } from '../projects/projects'
import type { PlayerSave } from '../save/types'
import { skillMenuPlacementForOutput, skillMenuView } from '../skills/skillActions'
import { recipeBookForSkill, type RecipeBookEntry } from './knowledge'

export interface RecipeBookSection {
  title: string | null
  entries: RecipeBookEntry[]
}

export interface RecipeBookTab {
  id: string
  label: string
  sections: RecipeBookSection[]
}

export interface RecipeBookView {
  skillId: string
  tabs: RecipeBookTab[]
}

export function recipeBookSkillId(
  db: GameDatabase,
  entries: RecipeBookEntry[],
): string | null {
  const first = entries[0]
  if (!first) return null
  if (first.kind === 'recipe') {
    const skillId = getRecipe(db, first.id)?.['Skill ID']
    return typeof skillId === 'string' && skillId ? skillId : null
  }
  const skillId = getProject(db, first.id)?.['Skill ID']
  return typeof skillId === 'string' && skillId ? skillId : null
}

export function recipeBookView(
  save: PlayerSave,
  db: GameDatabase,
  skillId: string,
): RecipeBookView {
  return recipeBookViewForEntries(db, skillId, recipeBookForSkill(save, db, skillId))
}

export function recipeBookViewForEntries(
  db: GameDatabase,
  skillId: string,
  entries: RecipeBookEntry[],
): RecipeBookView {
  const buckets = new Map<
    string,
    { id: string; label: string; sections: Map<string, RecipeBookEntry[]>; order: string[] }
  >()
  const sectionKey = (title: string | null) => title ?? ''
  for (const entry of entries) {
    const outputId =
      entry.kind === 'project'
        ? (getProject(db, entry.id)?.['Output Item / Target ID'] ?? '')
        : (getRecipe(db, entry.id)?.['Output Item ID'] ?? '')
    const placement = skillMenuPlacementForOutput(db, skillId, entry.name, outputId)
    let bucket = buckets.get(placement.tabId)
    if (!bucket) {
      bucket = { id: placement.tabId, label: placement.tabLabel, sections: new Map(), order: [] }
      buckets.set(placement.tabId, bucket)
    }
    const key = sectionKey(placement.sectionTitle)
    if (!bucket.sections.has(key)) {
      bucket.order.push(key)
      bucket.sections.set(key, [])
    }
    bucket.sections.get(key)!.push(entry)
  }
  const tabOrder = skillMenuView(db, skillId).tabs.map((tab) => tab.id)
  const toTab = (bucket: {
    id: string
    label: string
    sections: Map<string, RecipeBookEntry[]>
    order: string[]
  }): RecipeBookTab => ({
    id: bucket.id,
    label: bucket.label,
    sections: bucket.order.map((key) => ({
      title: key.length === 0 ? null : key,
      entries: bucket.sections.get(key)!,
    })),
  })
  const tabs: RecipeBookTab[] = [
    ...tabOrder.flatMap((id) => {
      const bucket = buckets.get(id)
      return bucket ? [toTab(bucket)] : []
    }),
    ...[...buckets.values()].flatMap((bucket) =>
      tabOrder.includes(bucket.id) ? [] : [toTab(bucket)],
    ),
  ]
  if (tabs.length === 0) {
    return { skillId, tabs: [{ id: 'actions', label: 'Actions', sections: [] }] }
  }
  return { skillId, tabs }
}

File: packages/ik_rules/lib/src/recipes/book_view.dart
import 'package:ik_content/ik_content.dart';

import '../js_compat.dart';
import '../production/recipes.dart';
import '../projects/projects.dart';
import '../save/generated/save_models.dart';
import '../skills/skill_actions.dart';
import 'knowledge.dart';

/// One group inside a recipe-book tab.
class RecipeBookSection {
  const RecipeBookSection({this.title, required this.entries});

  final String? title;
  final List<RecipeBookEntry> entries;
}

/// One tab of a recipe book, matching the skill-menu tab labels.
class RecipeBookTab {
  const RecipeBookTab({required this.id, required this.label, required this.sections});

  final String id;
  final String label;
  final List<RecipeBookSection> sections;
}

/// Tabbed recipe book for one skill or a station's filtered rows.
class RecipeBookView {
  const RecipeBookView({required this.skillId, required this.tabs});

  final String skillId;
  final List<RecipeBookTab> tabs;
}

String? recipeBookSkillId(GameDatabase db, List<RecipeBookEntry> entries) {
  if (entries.isEmpty) return null;
  final first = entries.first;
  if (first.kind == 'recipe') {
    final skillId = getRecipe(db, first.id)?.raw['Skill ID'];
    return skillId is String && skillId.isNotEmpty ? skillId : null;
  }
  final skillId = getProject(db, first.id)?.raw['Skill ID'];
  return skillId is String && skillId.isNotEmpty ? skillId : null;
}

RecipeBookView recipeBookView(PlayerSave save, GameDatabase db, String skillId) {
  return recipeBookViewForEntries(db, skillId, recipeBookForSkill(save, db, skillId));
}

RecipeBookView recipeBookViewForEntries(
  GameDatabase db,
  String skillId,
  List<RecipeBookEntry> entries,
) {
  final buckets = <String, _TabBucket>{};
  for (final entry in entries) {
    final outputId = entry.kind == 'project'
        ? jsString(getProject(db, entry.id)?.raw['Output Item / Target ID'])
        : jsString(getRecipe(db, entry.id)?.raw['Output Item ID']);
    final placement = skillMenuPlacementForOutput(db, skillId, entry.name, outputId);
    final bucket = buckets.putIfAbsent(
      placement.tabId,
      () => _TabBucket(id: placement.tabId, label: placement.tabLabel),
    );
    bucket.add(placement.sectionTitle, entry);
  }
  final tabOrder = [for (final tab in skillMenuView(db, skillId).tabs) tab.id];
  final tabs = <RecipeBookTab>[
    for (final id in tabOrder)
      if (buckets[id] case final bucket?) bucket.toTab(),
    for (final bucket in buckets.values)
      if (!tabOrder.contains(bucket.id)) bucket.toTab(),
  ];
  if (tabs.isEmpty) {
    return RecipeBookView(
      skillId: skillId,
      tabs: const [RecipeBookTab(id: 'actions', label: 'Actions', sections: [])],
    );
  }
  return RecipeBookView(skillId: skillId, tabs: tabs);
}

class _TabBucket {
  _TabBucket({required this.id, required this.label});

  final String id;
  final String label;
  final List<String?> _sectionOrder = <String?>[];
  final Map<String?, List<RecipeBookEntry>> _sections = <String?, List<RecipeBookEntry>>{};

  void add(String? title, RecipeBookEntry entry) {
    if (!_sections.containsKey(title)) {
      _sectionOrder.add(title);
      _sections[title] = <RecipeBookEntry>[];
    }
    _sections[title]!.add(entry);
  }

  RecipeBookTab toTab() {
    return RecipeBookTab(
      id: id,
      label: label,
      sections: [
        for (final title in _sectionOrder)
          RecipeBookSection(title: title, entries: _sections[title]!),
      ],
    );
  }
}
